examples/client/Locomotion/src/services/one-signal/index.ts

```diff
@@ -0,0 +1,171 @@
+      } as PushUserData);
+    }
+  };
+
+  getPushSettings = async (): Promise<PushSettings | null> => {
+    try {
+      const [
+        isPushEnabled,
+        pushToken,
+        pushSubscriptionId,
+      ] = await Promise.all([
+        OneSignal.User.pushSubscription.getOptedInAsync(),
+        OneSignal.User.pushSubscription.getTokenAsync(),
+        OneSignal.User.pushSubscription.getIdAsync(),
+      ]);
+
+      return {
+        isPushEnabled,
+        pushToken,
+        pushSubscriptionId,
+      };
+    } catch (error) {
+      console.error('Error getting device state', error);
+      return null;
+    }
+  };
+
+  refreshPushSettings = async (): Promise<PushSettings | null> => {
+    try {
+      const pushSettings = await this.getPushSettings();
+      if (!pushSettings) return null;
+
+      const {
+        isPushEnabled,
+        pushToken,
+        pushSubscriptionId,
+      } = pushSettings;
+
+      Mixpanel.setEvent('Notification Service: Check App State', { isSubscribed: isPushEnabled, pushToken, userId: pushSubscriptionId });
+
+      if (pushToken && pushSubscriptionId && isPushEnabled) {
+        await this.updateServer(pushToken, pushSubscriptionId, isPushEnabled);
+      }
+
+      return pushSettings;
+    } catch (error) {
+      console.error('Error checking latest device state', error);
+      return null;
+    }
+  };
+
+  handleNotificationClick = (openResult: NotificationClickEvent): void => {
+    const { notification } = openResult;
+    const additionalData = notification.additionalData as NotificationAdditionalData;
+
+    if (additionalData?.type && this.notificationsHandlers[additionalData.type]) {
+      this.notificationsHandlers[additionalData.type](additionalData);
+    }
+  };
+
+  handleForegroundNotificationClick = (
+    notificationReceivedEvent: NotificationWillDisplayEvent,
+  ): void => {
+    const { notification } = notificationReceivedEvent;
+    const additionalData = notification.additionalData as NotificationAdditionalData;
+
+    if (additionalData?.type && this.foregroundNotificationsHandlers[additionalData.type]) {
+      this.foregroundNotificationsHandlers[additionalData.type](additionalData);
+    }
+
+    notificationReceivedEvent.preventDefault();
+    notificationReceivedEvent.getNotification().display();
+  };
+
+  init = async (userId: string): Promise<PushSettings | null> => {
+    if (!Config.ONESIGNAL_APP_ID) return null;
+
+    OneSignal.initialize(Config.ONESIGNAL_APP_ID);
+    OneSignal.login(userId);
+
+    OneSignal.User.pushSubscription.optIn();
+
+    OneSignal.Notifications.addEventListener('click', this.handleNotificationClick);
+    OneSignal.Notifications.addEventListener('foregroundWillDisplay', this.handleForegroundNotificationClick);
+    OneSignal.User.pushSubscription.addEventListener('change', this.subscriptionObserverHandler);
+
+    if (Platform.OS === IOS) {
+      const permission = await OneSignal.Notifications.requestPermission(true);
```

Comment:
passing true will refer the user to open settings if they declined push notification permissions, see how we fixed it for the driver app:
https://github.com/Autofleet/driver-app/pull/1898

Reply:
Updated